src/OpenTelemetry/Metrics/MeterProviderSdk.cs

```diff
@@ -61,13 +63,20 @@ public class MeterProviderSdk
             this.listener = new MeterListener()
             {
                 InstrumentPublished = (instrument, listener) =>
+                {
+                    Console.WriteLine($"Instrument {instrument.Meter.Name}:{instrument.Name} published.");
+                    if (meterSourcesToSubscribe.ContainsKey(instrument.Meter.Name))
                     {
-                        Console.WriteLine($"Instrument {instrument.Meter.Name}:{instrument.Name} published.");
-                        if (meterSourcesToSubscribe.ContainsKey(instrument.Meter.Name))
+                        var instrumentState = new InstrumentState(this, instrument);
+
+                        lock (this.lockInstrumentStates)
                         {
-                            listener.EnableMeasurementEvents(instrument, null);
+                            this.instrumentStates.Add(instrument, instrumentState);
                         }
-                    },
+
+                        listener.EnableMeasurementEvents(instrument, instrumentState);
```

Comment:
Shouldn't we be disposing the `listener` in the `Dispose` method?

Reply:
Yes. Good catch!